api/api.go

```diff
@@ -0,0 +1,91 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/hashicorp/consul-terraform-sync/event"
+)
+
+const (
+	defaultAPIVersion = "v1"
+
+	statusHealthy      = "healthy"
+	statusDegraded     = "degraded"
+	statusCritical     = "critical"
+	statusUndetermined = "undetermined"
+)
+
+// API supports api requests to the cts biniary
+type API struct {
+	store   *event.Store
+	port    int
+	version string
+}
+
+// NewAPI create a new API object
+func NewAPI(store *event.Store, port int) *API {
+	return &API{
+		port:    port,
+		store:   store,
+		version: defaultAPIVersion,
+	}
+}
+
+// Serve starts up and handles shutdown for the http server to serve
+// API requests
+func (api *API) Serve(ctx context.Context) {
+	mux := http.NewServeMux()
+
+	// retrieve task status for a task-name
+	mux.Handle(fmt.Sprintf("/%s/%s/", defaultAPIVersion, taskStatusPath),
+		newTaskStatusHandler(api.store, api.version))
+	// retrieve all task statuses
+	mux.Handle(fmt.Sprintf("/%s/%s", defaultAPIVersion, taskStatusPath),
+		newTaskStatusHandler(api.store, api.version))
+
+	srv := &http.Server{
+		Addr:         fmt.Sprintf(":%d", api.port),
+		WriteTimeout: time.Second * 15,
+		ReadTimeout:  time.Second * 15,
+		IdleTimeout:  time.Second * 60,
+		Handler:      mux,
+	}
+
+	go func() {
+		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			// go ahead and stop cts from continuing
+			log.Fatalf("error starting api server at '%d': '%s'\n", api.port, err)
```

Comment:
I tried a couple of things here and couldn't find something that didn't start to get complicated for what I wanted to achieve. If people have ideas here, please share!

Reply:
Yeah let's look into updating this so it doesn't fatal mid-process and cause the other go routines (task execution / terraform runs) to stop abruptly.

Reply:
Thanks for pairing on this @findkim and talking to me about go-routines and context! I've made a pass at this as we talked about and tested that it works for a request that takes a long time to handle. Let me know what you think!

Reply:
Nice. The rework of the go-routine handling looks good!